shcomplete/seq2seq_correction.py

```diff
@@ -0,0 +1,225 @@
+
+    def decode(self, X, max_cmd_len, reduction=True, inverted=False):
+        """
+        Decode the numpy array X and return the corresponding command.
+        """
+        command = ""
+        if reduction:
+            X = X.argmax(axis=-1)
+            if inverted:
+                begin_of_command = np.amin(np.nonzero(X))
+                for i in range(begin_of_command, max_cmd_len):
+                    command += self.indices_char[X[i]]
+            else:
+                end_of_command = np.amax(np.nonzero(X))
+                for i in range(end_of_command + 1):
+                    command += self.indices_char[X[i]]
+            return command
+        elif len(np.nonzero(X)[0]) != 0:
+            end_of_command = np.amax(np.nonzero(X))
+            for i in range(end_of_command + 1):
+                command += self.indices_char[X[i]]
+        return command
+
+
+class OnEpochEndCallback(Callback):
+
+    def __init__(self, path_to_vocab, file_delimiter, path_to_corpus, models_directory, gConfig):
+        self.gConfig = gConfig
+        self.vocab = path_to_vocab
+        self.file_delimiter = file_delimiter
+        self.corpus = path_to_corpus
+        self.models_directory = models_directory
+
+    def on_epoch_end(self, epoch, logs=None):
+        """
+        Apply the prediction of the model to a batch of data at the end of each epoch.
+        """
+        chars = get_chars(self.vocab)
+        X, y = next(generator(chars, self.corpus, self.file_delimiter, self.vocab, self.gConfig))
+        sample_prediction(self.model, chars, X, y,
+                          self.gConfig["max_cmd_len"],
+                          self.gConfig["inverted"])
+        path_to_model = os.path.join(self.models_directory, "keras_spell_e{}.h5".format(epoch))
+        if epoch % 100 == 0:
+            self.model.save(path_to_model)
+
+
+def generator(chars, corpus, file_delimiter, path_to_vocab, gConfig):
+    """
+    An epoch finishes when  samples_per_epoch samples have been seen by the model.
+    """
+    seq2seq = Seq2seq(chars)
+    vocab = Vocabulary(path_to_vocab)
+    vocab_trie = vocab.trie(path_to_vocab)
+    commands = []
+    misspelled_commands = []
+    while True:
+        with open(corpus) as f:
+            lines = [cmd.strip() for cmd in f.readlines()
+                     if cmd != file_delimiter and len(cmd) <= gConfig["max_cmd_len"]]
+            for _ in range(gConfig["batch_size"]):
+                command = choice(lines)
+                misspelled_command, command = generate_input(chars, command, vocab_trie, gConfig)
+                commands.append(command)
+                assert len(command) == gConfig["max_cmd_len"]
+                if gConfig["inverted"]:
+                    misspelled_command = misspelled_command[::-1]
+                misspelled_commands.append(misspelled_command)
+            assert len(commands) == len(misspelled_commands)
+            assert len(commands) == gConfig["batch_size"]
```

Comment:
I see you access gConfig many times like that. Better to create an `argparse.Namespace` on the same dict and reference settings like that:  `gConfig.batch_size` This way you will be able to refactor easier